### Title: 
# Exploratory Data Analysis

### Description:
We are going to perform an EDA of our dataset.

### Authors:
#### Hugo Cesar Octavio del Sueldo
#### Jose Lopez Galdon

### Date:
23/10/2020

### Version:
1.0

***

### Libraries

We are going to start this EDA loading the respective libraries

In [1]:
    # Numpy & Pandas to work with the DF
import numpy as np
import pandas as pd

    # Seaborn / matplotlib for graphs
import seaborn as sns
import matplotlib.pyplot as plt 
%matplotlib inline

    # Visualize DF
from IPython.display import display, HTML

***

## Load data

In [15]:
    # To automate the work as much as possible, we will parameterize the codes, so in this case, we will create an objetct with
    # the path root
name = 'data_train_merge'

data = pd.read_csv(f'../data/02_intermediate/{name}.csv',  # Path root: here we include an f-string with the variable name to parametrize
                   low_memory = False)                     # To avoid warnings we use set low_memory = False

***

## Variable predict

Taking into account the goals of the practice, we only want to know those persons that `Fully paid` their loans or those who does not, so we will tansform de dependent variable into binary problem, to ones or ceros depending on that condition. 

In [16]:
    # We will only select those observations with "Fully Paid" & "Charged Off"
data_binary = data[(data['loan_status'] == "Fully Paid") | (data['loan_status'] == "Charged Off")]

    # Now, we will transform into 0 & 1
dummy_dict = {"Fully Paid":0, "Charged Off":1}

    # Finally, we use the dictiony in the dataset
data = data_binary.replace({"loan_status": dummy_dict})

### View data

Firstly, we are goint to take a look to our dataframe:

In [4]:
    # First 10 rows using html display in order to view all the columns
data = data
display(HTML(data.head(10).to_html()))

In [5]:
    # Last 10 rows using html display in order to view all the columns
data = data
display(HTML(data.tail(10).to_html()))

In [6]:
    # Data dimension
data.shape

(442567, 151)

### Data wrangling

Once we have viewed our dataset, we continue exploring the column types and nulls...

In [7]:
data.info(verbose = True, null_counts = True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 442567 entries, 8 to 884875
Data columns (total 151 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   id                                          442567 non-null  int64  
 1   member_id                                   0 non-null       float64
 2   loan_amnt                                   442567 non-null  int64  
 3   funded_amnt                                 442567 non-null  int64  
 4   funded_amnt_inv                             442567 non-null  float64
 5   term                                        442567 non-null  object 
 6   int_rate                                    442567 non-null  object 
 7   installment                                 442567 non-null  float64
 8   grade                                       442567 non-null  object 
 9   sub_grade                                   442567 non-null  object 


As we can see above we have plenty of columns with nulls, so we are going to drop those columns which more than a 75% of the information is null.

In [17]:
    # For this operation we will use drop.na(), but firstly we will create an objet with the scale
thresh = round(len(data) * 0.75) 

data = data.dropna(thresh = thresh, axis = 1)

    #Lets check the data shape
data.shape

We can see that now we have only 93 columns, instead of 151. This is a huge reduction of the variables, but we continue having too much, in order to reduce them, we will study which of them are important in order to predict `loan_status`.